ZEP-0008 Introduce `zarf dev inspect` and change `zarf package inspect` to a parent command (#9)
zarf-dev/proposals: 8-show-manifest-and-values/README.md

      - [Prerequisite testing updates](#prerequisite-testing-updates)
      - [Unit tests](#unit-tests)
      - [e2e tests](#e2e-tests)
  - [Graduation Criteria](#graduation-criteria)
  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
  - [Version Skew Strategy](#version-skew-strategy)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

<!--
This section is key for creating high-quality, user-focused documentation
like release notes or a roadmap. You should gather this info before
implementation starts to keep the focus on development, not writing. ZEP
editors should ensure the `Summary` is clear and useful for a broad audience.

A good summary should be at least a paragraph long.

Follow the [documentation style guide] for this section and the rest of the ZEP.
Keep line lengths reasonable to make it easier for reviewers to provide
feedback and reduce unnecessary changes.

[documentation style guide]: https://docs.zarf.dev/contribute/style-guide/
-->

Both creators and deployers need a way to view their manifests and values files after Zarf variables are applied.

> This will be accomplished through new CLI commands `zarf dev show-manifests`, `zarf dev show-values-files`, `zarf package show-manifests`, and `zarf package show-values-files`

Author: That's actually how I had it originally, but I ended up changing it as it felt like two different commands. Originally I had written out `zarf package preview --show-manifests` and `zarf package preview --show-values-files`. However we need to be able to show the manifests and values file for already built packages as well as on packages before they're built so we needed to add a `zarf package inspect --show-manifests` and `zarf package inspect --show-values-file` command as well, zarf package inspect is already an existing command.  

I will build out the alternatives more with these suggestions so we can reevaluate. There's a few different ways we can do it. I think sub commands will likely be the way to go. IMO the biggest challenge here is that things need to work with both packages definitions and already built packages and certain flags will only be applicable to one or the other.

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this ZEP.  Describe why the change is important and the benefits to users. You
can also optionally include links to [experience reports], [community slacks],
or other references to show the community's interest in the ZEP.

[experience reports]: https://go.dev/wiki/ExperienceReports
[openssf slack]: https://openssf.slack.com/archives/C07AKUMBDMJ
[kubernetes slack]: https://kubernetes.slack.com/archives/C03B6BJAUJ3
-->

> Viewing manifests and values files after Zarf variable templating would be useful for both creators and deployers. Catching a mistake in templating early can reduce cycle time. A Helm template is almost instant, whereas create + deploy could take several minutes to hours.

Author: Yeah we have a `zarf dev lint` command. I can't find the issue for it, but I think there's some enhancements we could add to lint that would help here. For example, we could make sure that any variable defined is used somewhere in the package / manifests. And now that I think about it we could probably also make sure that after templating everything is valid yaml. 

Still I think users would want a way to check what their manifests are, and if variables are being set in the correct way.

Before printing the manifest for each chart the name and version of the chart will be printed. Before printing Manifests from the `.components[x].manifests` key the name of the manifests block, `.components[x].manifests[x].name`, will be printed.

Below is the intended help text for `zarf dev show-manifests`. `zarf dev show-values-files` will include the same flags.
```
Usage:
  zarf dev show-manifests [ DIRECTORY ] [flags]

Flags:
      --create-set stringToString   Specify package variables to set on the command line (KEY=value) (default [])
      --deploy-set stringToString   Specify deployment variables to set on the command line (KEY=value) (default [])
  -f, --flavor string               The flavor of components to include in the resulting package (i.e. have a matching or empty "only.flavor" key)
      --kube-version                Override the default helm template KubeVersion when performing a package chart template

>       --confirm                     Confirms command without prompting. Skips prompts to configure variables.

Reviewer: Again, `--confirm` in a preview seems like an overkill.

Author: I think it's reasonable given that it would only prompt if there are variables that have the `prompt: true` key or templates that are not set (we error when templates aren't set)

```

Below is the intended help text for `zarf package show-manifests`. `zarf package show-values-files` will include the same flags.
```
Usage:
  zarf package show-manifests [ PACKAGE_SOURCE ] [flags]

Flags:
      --set stringToString          Specify deployment variables to set on the command line (KEY=value) (default [])
      --kube-version                Override the default helm template KubeVersion when performing a package chart template
      --components                  Comma-separated list of components whose manifests should be displayed.  Adding this flag will skip the prompts for selected components.  Globbing component names with '*' and deselecting 'default' components with a leading '-' are also supported

>       --confirm                     Confirms command without prompting. Skips prompts to configure variables and select optional components

Reviewer: Ditto.

```

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this ZEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

> As a creator of Zarf packages, I want to make sure the variables in my package can get templated properly for the expected values of the deployers. I want to check this for both manifests and values files so I run `zarf dev show-manifests path/to/package-dir --deploy-set=MY_VAR=my-val` and `zarf dev show-values-files path/to/package-dir --deploy-set=MY_VAR=my-val`

Reviewer: ```suggestion
As a creator of Zarf packages, I want to make sure the variables in my package are properly rendered with the expected values. I want to check this for both manifests and values files so I run `zarf dev show-manifests path/to/package-dir --deploy-set=MY_VAR=my-val` and `zarf dev show-values-files path/to/package-dir --deploy-set=MY_VAR=my-val`
```

Author: Good recs! I took them, but didn't use the commit feature because it messes with DCO :disappointed:

#### Story 2

> As a deployer of Zarf packages, I want to check that the variables I intend to deploy my package with are getting properly templated for both manifests and values files before I deploy so I run `zarf package show-manifests zarf-package-podinfo-amd64.tar.zst -set=MY_VAR=my-val` and `zarf package show-values-files zarf-package-podinfo-amd64.tar.zst --set=MY_VAR=my-val`

Reviewer: I don't see how this story is different from the first one, other than the first is using directory, and this a zst file.

Author: I wanted to highlight that this will be used both before a package is created, and after a package is created. I added the different flags that the different commands have to make the stories more clear.
